src/domain/core/src/entities/metadata_block_repository.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use opendatafabric::{MetadataBlock, Multihash};
+
+use crate::{
+    ContainsError,
+    GetBlockError,
+    GetError,
+    InsertError,
+    InsertOpts,
+    InsertResult,
+    ObjectRepository,
+};
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait]
+pub trait MetadataBlockRepository: Send + Sync {
+    async fn contains(&self, hash: &Multihash) -> Result<bool, ContainsError>;
+
+    async fn get_block(&self, hash: &Multihash) -> Result<MetadataBlock, GetBlockError>;
+
+    async fn get_block_data(&self, hash: &Multihash) -> Result<Bytes, GetError>;
```

Comment:
Seems unfortunate for this type to now work with two different levels of abstraction: blocks and bytes...

If we store bytes in the cache we might as well remove `MetadataBlockRepository` as such and go back to caching on `ObjectRepository` level, since the goal of adding `MetadataBlockRepository` was to avoid the deserialization overhead on cache hits.

@zaychenko-sergei perhaps transfer protocol should handle metadata blocks on a lower level:
- when iterating it could accumulate data as bytes
- when calculating sizes - it could create flatbuffer proxies over those bytes to read data directly without deserializing into DTOs (avoids lots of allocations)
- thus perhaps it could only scan metadata once, while maintaining its own "cache"

Reply:
1) Put it differently: cache could store both raw bytes and deserialized version simultaneously.  In addition, the deserialized version could also be lazily computed (thus never or at most once) to avoid ovearhead. It still keeps the new trait at the block abstraction level, and preserves effeciency for low-level access (bytes, size).

2) Protocol benefits from the implemented idea of dataset per-transaction cache. It's currently already doing 2 traversals:
   - at step 1 to create a size estimate
   - at step 2 to pack a tarball out of block bytes
As the `Dataset` object is kept between steps of the protocol, we can do just 1 set of reads with the cache involved. 

With 1 implemented as I describe, there is no deserialization overhead (lazy thing not computed).

3) The protocol should rather be kept isolated from storage/caching details, as it currently is. I would only go into custom option if we could sum sizes / zip stream for an interval range more effeciently than doing it individually in a loop. 

If we are fantasizing here, block could have a field meaning a sum of bytes in all the previouls blocks, thus size estimate could be computed  as a difference between these numbers in 2 selected blocks making a range.

Reply:
> 1. Put it differently: cache could store both raw bytes and deserialized version simultaneously. In addition, the deserialized version could also be lazily computed (thus never or at most once) to avoid ovearhead. It still keeps the new trait at the block abstraction level, and preserves effeciency for low-level access (bytes, size).

Implemented